shuffle sharding package for priority and fairness

mars1024 opened the pull request:
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Implement package for shuffle sharding algorithm, which will be used in KEP [priority and fariness](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md) .

**Special notes for your reviewer**:

Squashed PRs #80710 #81182 , related PRs #81833 (need to be merged in flowcontrol package later)

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

- KEP: [https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md)

mars1024 committed:
Add shuffle sharding utils and tests

Implement several shuffle sharding functions including ShuffleAndDeal,
ShuffleAndDealToSlice.

Add benchmarks and tests for shuffle sharding to test performance,
correctness and distribution uniformity.

Signed-off-by: Bruce Ma <brucema19901024@gmail.com>

MikeSpreitzer committed:
More work on shuffle sharding utils

Changes following up on shuffle sharding util package.

Made the validation checking function return a slice of error messages
rather than just a bit.

Replaced all the `int32` with `int` because this is intended for more
than just the priority-and-faireness feature and so should not be a
slave to its configuration datatypes.

Introduced ShuffleAndDealIntoHand, to make memory allocation the
caller's problem/privilege.

Made the hand uniformity tester avoid reflection, evaluate the
histogram against the expected range of counts, and run multiple test
cases, including one in which the number of hash values is a power of
two with four extra bits (as the validation check requires) and one in
which the deck size is not a power of two.

mars1024 commented:
/cc @yue9944882 @MikeSpreitzer @aaron-prindle @deads2k @lavalamp @yliaog

mars1024 commented:
/retest

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go:
```diff
@@ -0,0 +1,133 @@
+// sharding.  Returns a slice for which `len()` is 0 if and only if
+// there are no errors.  The entropy requirement is evaluated in a
+// fast but approximate way: bits(deckSize^handSize).
+func ValidateParameters(deckSize, handSize int) (errs []string) {
```
This function was intended to cover all the restrictions in https://github.com/kubernetes/kubernetes/blob/feature-rate-limiting/pkg/apis/flowcontrol/validation/validation.go#L221-L233 .  But it is actually only used for one test in there, because the list returned from this function is not adequately expressive to drive all those validation tests.  It is awkward for https://github.com/kubernetes/kubernetes/blob/feature-rate-limiting/pkg/apis/flowcontrol/validation/validation.go#L231 to invoke this whole function just to get one test.  I suggest that we either (a) change the returned value so that this function can do all of the testing in https://github.com/kubernetes/kubernetes/blob/feature-rate-limiting/pkg/apis/flowcontrol/validation/validation.go#L221-L233 or (b) factor out of this function the one test that the validation code needs.

yue9944882 replied:
(b) sounds probably better. we factor a `RequiredEntropyBits` out of this and make the new function returning one error.

mars1024 replied:
will factor this out, let do the most validations in flowcontrol.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go:
```diff
@@ -0,0 +1,133 @@
+// ShuffleAndDeal can shuffle a hash value to handSize-quantity and non-redundant
+// indices of decks, with the pick function, we can get the optimal deck index
+// Eg. From deckSize=128, handSize=8, we can get an index array [12 14 73 18 119 51 117 26],
+// then pick function will choose the optimal index from these
```
This comment reflects an incomplete transition from the original context of fair queueing to a general-purpose utility.  I recommend completing the transition.  The word "index" is used here, but for general purpose usage we should follow the metaphor of shuffle-sharding.  This is about a "deck" of "cards".  Thus, the comment should be more like the following.

> ShuffleAndDeal shuffles a card deck of the given size and deals a hand of the given size, using the
> given hashValue as a source of entropy.
> This function makes handSize sequential calls to pick, one for each card.
> Each card is identified by an integer in the range [0, deckSize).
> For example, for deckSize=128 and handSize=4 this function might call pick(14); pick(73); pick(119); pick(26).

MikeSpreitzer replied:
The use in fair queuing does not call the one that returns a slice.
Defining a function to return a slice is the worst way to handle memory.
Better to have the caller allocate the memory.
Better still, if you easily can, to not involve a slice at all.

lavalamp replied:
OK, if I imagine a little harder I can see how the function can work if you just want to pick the "best" one in some sense.

Less memory allocs is usually better and it is easy to let the caller preallocate, so it's fine to accept a slice as well as construct one (I hadn't seen that version yet when I made this comment). It's hard for me to call this important yet without profiling the system as a whole, I'm optimizing for readability at this stage.

(My guess, not that we're at the stage to be optimizing for this yet, is that caller providing a pre-allocated slice is fastest, followed by providing a pick() function, followed by allocating a slice.)

MikeSpreitzer replied:
I think having the fair queuing client pass a `pick` function makes for the most direct source code, so I am fine with it being that way now.

yue9944882 replied:
+1 valuing readability over memory-allocs here. we can refactor the pick function to `func(cards []int)` and replacing the word "index" with "card".

MikeSpreitzer replied:
The most direct code is what's there now.  Adding a slice of ints anywhere makes the code more complicated: the slice has to be allocated, the slice has to be populated, and the consumer has to loop over the members of the slice.  The current code is simpler: the implementation of `pick` is the body of that loop, and all the other overhead does not exist.

mars1024 replied:
Why pre-allocated slice is fastest here? I think we can also use pre-allocated slice in pick function. So we should keep pick(int) here? @MikeSpreitzer @yue9944882 @lavalamp

mars1024 replied:
> Mike's comment suggestion is a large improvement.
> 
> Is there a reason for this function to be public? I am having a hard time imagining when I would call this and not the one that gives me the slice.

I think this function is the basic function, which means that others can extend this function by using different pick functions, so I think it's necessary to keep this public.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go:
We do not need to discuss the implementation in this comment.  And we do not need unnecessary multiples. I suggest something more like the following.

> ShuffelAndDeadltoSlice will shuffle a deck of the given size and return a hand of the given size

mars1024 replied:
will remove this function.

MikeSpreitzer reviewed (commented):
I made some independent comments in-line.
Also, why not have this PR put the files in the proper directory?

lavalamp commented on staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go:
I would not phrase the problem this way (i.e., with validation on/off versions).

I'd make the public interface this if it were me:

```
func NewDealer(deckSize, handSize int) (*Dealer, error) {
  // validate & return
}

type Dealer struct {
  deckSize, handSize int
}

func (d *Dealer) Deal(hash uint64) (deck []int) {
  d.deal(hash, func(i int) {deck = append(deck, i)})
  return deck
}

```

MikeSpreitzer replied:
The API Priority and Fairness validation code has no need of a Dealer, it only needs to check validiity of the parameters.

lavalamp replied:
The request path code needs a dealer.

lavalamp replied:
My main point here is to separate initialization & verification from execution-- this prevents people from executing with invalid parameters while not paying the cost of verification on every execution.

MikeSpreitzer replied:
My complaint is that if validation is inseparably bound with execution then we will have either (a) two verification paths, the one that is bound up with execution and the one that is not (the one that the API Priority and Fairness validation code calls), or (b) the API Priority and Fairness validation code has to go through at least part of the rigamarole for execution and the validation complaints out of that need to be expressed in a way that API validation code can consume (which is not true of the Dealer proposal above).

I think we may be working this too hard.  I think it would be fine to just have separate validation and execution code.

mars1024 replied:
If shuffle sharding is only used in flowcontrol, I don't think we need a dealer, because ...
1. validation on shuffle sharding has been parted in flowcontrol and shufflesharding package
2. flowcontrol will use shuffle sharding properly

lavalamp commented on staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go:
I definitely don't think it's a good idea to double the number of functions by providing with and without validation versions of them.

MikeSpreitzer replied:
I would not be sorry to see these combinations deleted.  We do not need them for the API Priority and Fairness feature.

mars1024 replied:
Yeah, I will simplify the useless codes.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding_test.go:
```diff
@@ -0,0 +1,347 @@
+	}
+}
+
+func BenchmarkValidateParameters(b *testing.B) {
```
Why is this benchmarked? There is no reason for it to be in the serving path.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding_test.go:
```diff
@@ -0,0 +1,347 @@
+	}
+}
+
+func TestShuffleAndDealToSliceWithValidation(t *testing.T) {
```
There's no reason to validate every time you deal a hand.

lavalamp reviewed (commented):
First set of comments, I might have more but I think this public interface needs to be changed first.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go:
```diff
@@ -0,0 +1,133 @@
+	"strings"
+)
+
+const maxHashBits = 60
```
What's the client usage of this item?  If it is validation --- which is the only bit of validation that the API Priority and Fairness validation code delegates to here --- then perhaps we do not need the validation functionality here?

mars1024 replied:
Maybe we should keep this simpler (only keep entropy bits validation) instead of removing this?

mars1024 commented:
> I made some independent comments in-line.
> Also, why not have this PR put the files in the proper directory?

Which proper directory you mean?

mars1024 commented on staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go:
```diff
@@ -0,0 +1,133 @@
+	if handSize > deckSize {
+		return []string{"handSize is greater than deckSize"}
+	}
+	if math.Log2(float64(deckSize))*float64(handSize) > maxHashBits {
```
Thanks, will update this.

mars1024 committed:
address some comments

Clean up useless functions, only keep the basic function Deal
and the function DealIntoHand which will be used by Priority
and Fairness.

Improve some comments for constants and functions.

Introduce Dealer to combine parameters and methods into a whole.

Use fixed-size slice to improve performance.

Use math.Ceil and math.Log2 to calculate required entropy bits.

Make the given hand adaptive to handSize in DealIntoHand.

Signed-off-by: Bruce Ma <brucema19901024@gmail.com>

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go:
```diff
@@ -0,0 +1,78 @@
+// is more than MaxHashBits. The entropy requirement is evaluated in a
+// fast but approximate way: bits(deckSize^handSize).
+// This must be called before using ShuffleAndDeal.
+func RequiredEntropyBits(deckSize, handSize int) error {
```
The name of this function suggests that it returns a number of bits, not an error.  We should fix this mismatch.  There are a few ways we could do this.  One would be to change the return value to a number of bits.  That would be consistent with exposing the limit (MaxHashBits).  These two changes make it the client's responsibility to compare these two quantities and react appropriately.  Alternatively, we could leave the return from this function an `error` and change the name to something like `ValidateEntropyBits`.  But note, the string in the returned error is not exactly what we want in the API validation code.  Perhaps the best option is to make this function simply return a boolean indicating constraint satisfaction or violation, and let the client react as it likes.

lavalamp replied:
I was expecting an int.

mars1024 commented:
Updated, introduce `Dealer` and address some comments.

mars1024 commented:
/retest

mars1024 commented:
/retest

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go:
```diff
@@ -0,0 +1,104 @@
+// deckSize and handSize, and this value must be less or equal than MaxHashBits
+// if we want to create and use Dealer.
+func RequiredEntropyBits(deckSize, handSize int) int {
+	return int(math.Ceil(math.Log2(float64(deckSize)) * float64(handSize)))
```
Remember that floating point arithmetic has limited precision.  `math.Log2(x)` produces only an approximation to the true base-2 logarithm of `x`.  I see no guarantee that this approximation is `<=` the true value.  Thus, the approximation could be slightly higher than the true value.  `math.Ceil` of something slightly higher than an integer is almost a whole integer larger.  That could be a problem here.  I suggest
```
 	return int(math.Ceil(math.Log2(float64(deckSize)) * float64(handSize) - 0.001))
```

lavalamp replied:
https://play.golang.org/p/cdiupEIE4EN

It seems it actually has the opposite problem.

Is this a good way to test for this? Would computing a max number be better?

MikeSpreitzer replied:
That is, that test observed approximations that were low but none that were high.
Looking at the doc on the math package, I see no guarantee in either direction.
The formulation I suggested is safe for our purposes, both for approximations that are a little below the correct value and approximations that are a little above the correct value.

mars1024 replied:
@lavalamp @MikeSpreitzer  Thanks for your suggestion, according to https://play.golang.org/p/muwuKH3EWDL, maybe we can use `math/big`, inspiration from @yue9944882 ...

MikeSpreitzer replied:
Even `math/big` is approximate when dealing with real numbers.  So switching to that changes nothing.

Look carefully at the revision I suggested.  It works when the approximation is low, and it works with the approximation is high.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go:
```diff
@@ -0,0 +1,104 @@
+	deckSize int
+	handSize int
+
+	// hand and rawCards are pre-allocated to improve performance
```
Putting them here makes the Dealer non-thread-safe.  That is, it can not be used concurrently by multiple threads.  That is a problem.  We want a Dealer to be usable by multiple threads concurrently.

lavalamp replied:
+1 these can't be here. I don't even think you can use e.g. a sync.Pool to reuse, since you don't know how long a caller will use them for.

mars1024 replied:
@MikeSpreitzer @lavalamp Agree on the thread-safe on Dealer, but still have a question, since `hand` will be used by caller **after** shuffle process and `rawCards` is only used **in** shuffle process, could we use sync.Pool for `rawCards` and allocate int slice for `hand` every time?

MikeSpreitzer replied:
No.  Make it work like it does in the feature branch.  The lack of escape of the pointers should be easy for the compiler to figure out, so it can produce efficient code.  That way we have simple, direct source code and good emitted code.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go:
```diff
@@ -0,0 +1,104 @@
+		d.hand[idx] = int(card)
+		idx++
+	})
+	return d.hand
```
Dealer needs to be thread-safe.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go:
```diff
@@ -0,0 +1,115 @@
+// This function can be extended with different pick functions.
+func (d *Dealer) Deal(hashValue uint64, pick func(int)) {
+	rawCards := d.rawCardsPool.Get().([]int)
+	defer d.rawCardsPool.Put(rawCards)
```
There is a much simpler and more efficient way to do this:
```
    var rawCards [15]int
```

lavalamp replied:
+1, I would not use the sync.Pool for this.

mars1024 replied:
@MikeSpreitzer @lavalamp  I'm sorry I don't think I've got your points, sync.Pool will get us out of repeated allocation for rawCards slice in different threads... Why allocating rawCards slice every time is a simper and more efficient way?

MikeSpreitzer replied:
I am suggesting a fixed-size array, not a slice.  The fixed-size array will be allocated on the heap.

MikeSpreitzer replied:
Also, BTW, I think "remainders" was a better than for this variable than "rawCards".

mars1024 replied:
Got your point, will update soon ~

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go:
```diff
@@ -0,0 +1,115 @@
+// RequiredEntropyBits must be less or equal than MaxHashBits
+// if we want to create and use Dealer.
+func RequiredEntropyBits(deckSize, handSize int) int {
+	return big.NewInt(1).Exp(big.NewInt(int64(deckSize)), big.NewInt(int64(handSize)), nil).BitLen()
```
We do not need `big` here, and there is an off-by-1 error here.  Following is a much simpler and more efficient way to do this.
```
import "math/bits"
...
    return bits.Len64(deckSize-1)*handSize
```

lavalamp replied:
I agree we shouldn't need `big`, but I think RequiredEntropyBits should defend against overflow (e.g, handSize musn't be larger than MaxInt / 64 in Mike's version above - which is ridiculously large, arbitrarily choosing 100 or 50 as the limit is fine. Negative values should also result in failing - which in this case means returning MaxInt as the required entropy bits -- I think that's easiest.

MikeSpreitzer replied:
Oops, my suggestion has to be modified to get the types right.  It should be
```
import "math/bits"
...

func NewDealer(deckSize, handSize int) (*Dealer, error) {
...
        if deckSize > 1 << 26 {
                return nil, fmt.Errorf("deckSize %d is impractically large", deckSize)
        }
...
}
...


func RequiredEntropyBits(deckSize, handSize int) int {
    return bits.Len(uint(deckSize-1))*handSize
}
```

@lavalamp : do you realize that the other checks in the NewDealer constructor, together with the one I suggested adding, ensure that there is no overflow or confusion due to negative numbers?  Note that my suggested implementation of RequiredEntropyBits will not panic with any parameter values.  Note also that returning MaxInt is not exactly a clear way to say that the parameters were invalid; it is a hack that works in the intended context, which is the tests in the NewDealer constructor or any similar code.  As long as we are presuming the intended context, I think it is sufficient to document the parameter restrictions in a comment.  Following is a draft of such a comment.

```
// RequiredEntropyBits makes a conservative estimate of the number of bits of
// hash value that are consumed in shuffle sharding a deck of the given size to
// a hand of the given size.  The result is meaningful only if
// 1 <= handSize <= deckSize <= 1<<26.
```

MikeSpreitzer replied:
The 26th power of 2 is the highest one with the property that when multiplied by its base-2 logarithm the result is guaranteed to fit into a 32-bit signed integer without overflow.  https://golang.org/ref/spec#Numeric_types says that `int` is a 32 or 64 bit integer.

MikeSpreitzer replied:
If someone wants to argue that we must repeat the parameter checks inside RequiredEntropyBits --- making for three copies of those checks in our code --- then I think this is an argument that we are doing something wrong.  If we can not extract that function from its context without internalizing a copy of that context then, well, I think that is an argument that this function simply can not be extracted from its context.  If that is to be argued then I think the best course of action is to eliminate the RequiredEntropyBits function and put the required expression directly in the checking code in the constructor of Dealer.

mars1024 replied:
@lavalamp @MikeSpreitzer  I don't think we have reached a solution about where should we put and how should we use the validations in shuffle sharding package... 
I agree on the limited max value for deckSize from @lavalamp , 1<<26 seems too large for validation, maybe 1024 or 512 is more proper here?
And I don't think the off-by-1 error will stop us from using `bigint`, because bitsof(deckSize-1)*handSize is less accurate as bitsof(deckSize^handSize-1),  https://play.golang.org/p/5C9rB_0RxMT ,  in some cases, the difference between the two algorithms is 7 (36-29), I think this is unacceptable.

MikeSpreitzer replied:
Oh, right.  I was focusing only on deckSize values that are a power of two.  The estimation formula I suggested can indeed err in other cases.

There _is_ indeed an off-by-one error in the PR's current formulation using `math/big`.

See https://gist.github.com/MikeSpreitzer/206d449cb23e74fe8486895e483865ff for a comparison of three techniques for extimatimg the number of entropy bits:
```
big.NewInt(1).Sub(big.NewInt(1).Exp(big.NewInt(int64(deckSize)), big.NewInt(int64(handSize)), nil), big.NewInt(1)).BitLen()
```
,
```
int(math.Ceil( math.Log2(float64(deckSize)) * float64(handSize) ))
```
, and
```
int(math.Ceil( math.Log2(float64(deckSize)) * float64(handSize) - 1e-11))
```

It shows that all three produce identical results for 2 <= deckSize <= 1<<30 and result <= 60.

I did a little investigation of whether we could rely on `math.Log2` of a power of 2 producing an exactly correct answer.  I suspect that maybe we can.  Section 4.3 of IEEE Std 754-2008 (both as of 2008 and in the 2019 revision) says

> Except where stated otherwise, every operation shall be performed as if it first produced an intermediate result correct to infinite precision and with unbounded range, and then rounded that result according to one of the attributes in this clause.

The attributes control how rounding is done, but rounding is moot when the infinite precision result is an integer.

When deckSize is not a power of 2, we need not worry about the fact that floating point numbers are in general approximations because the error of the approximation is certain to be too small to make a difference in the particular cases we care about.

Thus, I favor the simplest code, to wit:
```
import "math"
...

// RequiredEntropyBits makes a quick and slightly conservative estimate of the number
// of bits of hash value that are consumed in shuffle sharding a deck of the given size
// to a hand of the given size.  The result is meaningful only if
// 1 <= handSize <= deckSize .
func RequiredEntropyBits(deckSize, handSize int) int {
    return int(math.Ceil( math.Log2(float64(deckSize)) * float64(handSize) ))
}
```

MikeSpreitzer replied:
I asked the Sophers math channel on Slack; the conversation _ended_ at https://gophers.slack.com/archives/CC84BTU6L/p1571278799066200 .  The answer is that Go's `math` package does _not_ generally "round correctly" according to IEEE Std 754-2008 and 2019.  However, we got lucky: we do not care about generalities, all we really need to know is whether `math.Log2` of `1<<k` produces exactly `k`.  And it does!  The source code is at https://golang.org/src/math/log10.go

So this confirms to me that my latest suggestion can be expected to work.

mars1024 replied:
I tried in https://play.golang.org/p/cO4_6D5XtKf and got the same result as @MikeSpreitzer , I will abandon math/big ...

mars1024 replied:
Updated as your suggestion.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go:
```diff
@@ -0,0 +1,115 @@
+
+// DealIntoHand shuffles a deck of the given size by the
+// given hash value and deals cards into a hand(slice).
+func (d *Dealer) DealIntoHand(hashValue uint64) []int {
```
Remember https://github.com/kubernetes/kubernetes/blob/feature-rate-limiting/staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go#L108

DealIntoHand should take the slice as a parameter, not return the slice.

lavalamp replied:
+1

mars1024 replied:
Thanks for catching this, I'll update this together after we come to a conclusion about the validations and rawCards above...

lavalamp commented:
I think we are super close now, thanks for bearing with us :)

mars1024 commented:
Updated! @lavalamp @MikeSpreitzer

mars1024 commented on staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go:
```diff
@@ -0,0 +1,106 @@
+
+// DealIntoHand shuffles a deck of the given size by the given hash value and
+// deals cards into a given hand(slice) which should be pre-allocated.
+// The size of the given hand must be equal to the initial handSize of the Dealer.
```
Last question, should we handle the situation that the size of given hand is not equal to the handSize of Dealer, do nothing or return error? Or just put this limitation in comments to let user obey this? @lavalamp  @MikeSpreitzer

MikeSpreitzer replied:
This function should return an `error` that is non-nil in that case and that case only.

lavalamp replied:
An alternative is:

```
func (d *Dealer) DealIntoHand(hashValue uint64, hand []int) []int {
	h := hand[:0]
	d.Deal(hashValue, func(card int) { h = append(h, int(card)) })
	return h
}
```

This has no chance of failing and is zero-alloc if a hand of the correct size is passed in. Callers don't have to check an error.

MikeSpreitzer replied:
Ah, you mean "a hand of an adequate size", right?  The comment should be updated too.

mars1024 replied:
Good idea from @lavalamp , thanks!

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go:
```diff
@@ -0,0 +1,105 @@
+	if handSize > deckSize {
+		return nil, fmt.Errorf("handSize %d is greater than deckSize %d", handSize, deckSize)
+	}
+	if deckSize > 1<<26 {
```
Since the function's comment is documenting error conditions, this one has to be included.

mars1024 replied:
done!

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go:
```diff
@@ -0,0 +1,105 @@
+// DealIntoHand shuffles a deck of the given size by the given hash value and
+// deals cards into a given hand(slice) which should be pre-allocated, if the
+// size of the given hand is not equal to the handSize of the Dealer, the given
+// hand will be expanded or shrunk to the "handSize", then will be returned.
```
This comment needs some adjusting.  The given slice is passed by value, so the caller sees no modification to that.  Exact equality is not important here, note that what we actually have is a one-sided constraint.  I suggest a revision like the following.

> // DealIntoHand suffles and deals according to the Dealer's parameters,
> // using the given hashValue as the source of entropy and
> // returning the dealt cards as a slice of `int`.  The returned slice uses the
> // backing array of the given slice if it is big enough --- which it should be;
> // the function will perform correctly but not optimally otherwise.

lavalamp replied:
Perhaps even clearer:

```
// DealIntoHand suffles and deals according to the Dealer's parameters,
// using the given hashValue as the source of entropy and
// returning the dealt cards as a slice of `int`.
// If `hand` is the correct length, it will be used as-is and no allocations will be made.
// If `hand` is nil or too small, it will be extended (performing an allocation).
// If `hand` is too large, a sub-slice will be returned.
```

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go:
```diff
@@ -0,0 +1,105 @@
+// This function makes handSize sequential calls to pick, one for each dealt card.
+// Each card is identified by an integer in the range [0, deckSize).
+// For example, for deckSize=128 and handSize=4 this function might call pick(14); pick(73); pick(119); pick(26).
+// This function can be extended with different pick functions.
```
This comment needs adjusting.  The parameters are no longer given here.  The statement that "This function can be extended with different pick functions" is awkward: "extend" is not quite right, the caller _must always_ supply a pick function, and it's obvious (since we say nothing to forbid it) that the caller can supply whatever pick function it wants.  I suggest something like the following.

> // Deal shuffles a card deck and deals a hand of cards, using the given hashValue as the source of entropy.
> // The deck size and hand size are properties of the Dealer.
> // This function synchronously makes sequential calls to pick, one for each dealt card.
> // Each card is identified by an integer in the range [0, deckSize).
> // For example, for deckSize=128 and handSize=4 this function might call pick(14); pick(73); pick(119); pick(26).

mars1024 replied:
Seems better, thanks!

MikeSpreitzer commented:
/retest

MikeSpreitzer commented:
FYI, I wondered whether returning the slice would confuse the compiler's escape analysis.  To investigate that, I ran the unit tests with `go test -gcflags -m .`.  That produced many remarks from the escape analysis, but none said that the given hand escapes in the unit test lines that invoke `Dealer::DealIntoHand`.

MikeSpreitzer commented:
Looking more carefully at the escape analysis remarks, I see the following.  The unit test file makes three calls to DealIntoHand.  The last two are coded in such a way that the given hand escapes to the heap regardless of what DealIntoHand does, and the analysis says that the hand escapes.  The first one is not coded in that way, but the given hand is always empty (`make([]int, 0)`).  For this the analysis says that it does not escape.

MikeSpreitzer commented:
I extended the last test function with some code that uses a non-empty given hand, and the analysis said it does not leak.  Here is the code:
```
	hf := make([]int, 6, 12)
	hf2 := dealer.DealIntoHand(42, hf)
	if len(hf2) != 6 {
		t.Errorf("Failed at last!")
	}
```

and here is what the analysis said:

```
./shufflesharding_test.go:314:12: TestDealer_DealIntoHand make([]int, 6, 12) does not escape
```

lavalamp commented:
I am satisfied by this, I think everything remaining is a nit.

/approve

(anyone can LGTM one you fix the comments)

Thanks for the escape analysis, Mike!

mars1024 commented:
> I extended the last test function with some code that uses a non-empty given hand, and the analysis said it does not leak. Here is the code:
> 
> ```
> 	hf := make([]int, 6, 12)
> 	hf2 := dealer.DealIntoHand(42, hf)
> 	if len(hf2) != 6 {
> 		t.Errorf("Failed at last!")
> 	}
> ```
> 
> and here is what the analysis said:
> 
> ```
> ./shufflesharding_test.go:314:12: TestDealer_DealIntoHand make([]int, 6, 12) does not escape
> ```

I've already add some tests for this.

mars1024 commented:
/retest

mars1024 commented:
/retest

yue9944882 commented:
/retest

yue9944882 commented:
/retest

MikeSpreitzer reviewed (approved):
LGTM

MikeSpreitzer commented:
/LGTM
